monitoring: add mount_point_info Prometheus metric

ggilmore opened the pull request:
This PR implements a new Prometheus metric named `mount_point_info`, which allows Zoekt instances to advertise the name of devices that are attached to them. 

See [the `MustRegisterNewMountPointInfoMetric` doc comment](https://github.com/sourcegraph/zoekt/blob/zoekt-disk-metrics/internal/mountinfo/mountinfo.go) for more information. 

Grafana screenshot of an example for `mount_point_info`. 
<img width="1650" alt="Screen Shot 2022-10-06 at 4 21 00 PM" src="https://user-images.githubusercontent.com/9022011/194435705-e3444744-0bd6-40ab-bf38-dd1c5ae99e37.png">

`mount_point_info` can be [joined](https://grafana.com/blog/2021/08/04/how-to-use-promql-joins-for-more-effective-queries-of-prometheus-metrics-at-scale/) on disk metrics that are emitted from [node_exporter](https://github.com/prometheus/node_exporter), which allows us to write queries that show us stats for the _specific disks_ that are attached to Zoekt instances. 

Example of the Grafana metrics that we can now create: reads/sec for the specific disk that's attached to zoekt websever 

```promql
mount_point_info{mount_name="indexDir", instance="zoekt-webserver-0:6070"} * on(device, node_name) group_left rate(node_disk_reads_completed_total[5m])
```
<img width="1658" alt="Screen Shot 2022-10-06 at 4 28 40 PM" src="https://user-images.githubusercontent.com/9022011/194436401-67987e82-2703-4897-89e4-9dfc1da2ea94.png">


# How to test (via docker-compose)

1. Use this script to build local versions of the zoekt-webserver and zoekt-indexserver docker images
	```bash
	#!/usr/bin/env bash
	
	cd "$(dirname "${BASH_SOURCE[0]}")"
	set -euxo pipefail 
	
	docker build -f Dockerfile -t zoekt --platform=linux/amd64 .
	docker build -f Dockerfile.webserver -t zoekt-webserver-local --platform=linux/amd64 .
	docker build -f Dockerfile.indexserver -t zoekt-indexserver-local --platform=linux/amd64 .
	```
1. Using the latest version of https://github.com/sourcegraph/deploy-sourcegraph-docker as a base, modify the following:
	```diff
	diff --git a/docker-compose/dev/docker-compose.yaml b/docker-compose/dev/docker-compose.yaml
	index 5e76915..c6fe5e1 100644
	--- a/docker-compose/dev/docker-compose.yaml
	+++ b/docker-compose/dev/docker-compose.yaml
	@@ -4,9 +4,34 @@
	 version: '2.4'
	 services:
	   zoekt-webserver-0:
	+    image: zoekt-webserver-local
	+    environment:
	+      - SRC_LOG_FORMAT=condensed
	+      - SRC_LOG_LEVEL=dbug
	     cpus: 1
	   zoekt-indexserver-0:
	+    image: zoekt-indexserver-local
	+    environment:
	+      - SRC_LOG_FORMAT=condensed
	+      - SRC_LOG_LEVEL=dbug
	     cpus: 1
	+  node-exporter:
	+    image: us.gcr.io/sourcegraph-dev/node-exporter:v1.4.0
	+    container_name: node-exporter
	+    restart: unless-stopped
	+    volumes:
	+      - /proc:/host/proc:ro
	+      - /sys:/host/sys:ro
	+      - /:/rootfs:ro
	+    command:
	+      - '--path.procfs=/host/proc'
	+      - '--path.rootfs=/rootfs'
	+      - '--path.sysfs=/host/sys'
	+      - '--collector.filesystem.mount-points-exclude=^/(sys|proc|dev|host|etc)($$|/)'
	+    expose:
	+      - 9100
	+    networks:
	+      - sourcegraph
	   caddy:
	     ports:
	       - '0.0.0.0:8080:80'
	diff --git a/prometheus/prometheus_targets.yml b/prometheus/prometheus_targets.yml
	index fae9d60..8ab1beb 100644
	--- a/prometheus/prometheus_targets.yml
	+++ b/prometheus/prometheus_targets.yml
	@@ -1,75 +1,98 @@
	 ---
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: node
	   targets:
	     - cadvisor:8080
	     - sourcegraph-frontend-internal:6060
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: github-proxy
	   targets:
	     - github-proxy:6060
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: repo-updater
	   targets:
	     - repo-updater:6060
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: worker
	   targets:
	     - worker:6060
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: worker-executors
	   targets:
	     - worker:6996
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: node
	   targets:
	     - syntect-server:6060
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: precise-code-intel-worker
	   targets:
	     - precise-code-intel-worker:6060
	     # Add new entries here for every searcher/symbol/gitserver replica.
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: zoekt-indexserver
	   targets:
	     - zoekt-indexserver-0:6072
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: zoekt-webserver
	   targets:
	     - zoekt-webserver-0:6070
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: sourcegraph-frontend
	   targets:
	     - sourcegraph-frontend-0:6060
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: gitserver
	   targets:
	     - gitserver-0:6060
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: searcher
	   targets:
	     - searcher-0:6060
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: symbols
	   targets:
	     - symbols-0:6060
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: pgsql
	   targets:
	     - pgsql-exporter:9187
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: codeintel-db
	   targets:
	     - codeintel-db-exporter:9187
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: codeinsights-db
	   targets:
	     - codeinsights-db-exporter:9187
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: redis-cache
	   targets:
	     - redis-cache:9121
	 - labels:
	+    nodename: "sourcegraph-docker-compose-host"
	     job: redis-store
	   targets:
	     - redis-store:9121
	+- labels:
	+    nodename: "sourcegraph-docker-compose-host"
	+    job: node-exporter
	+  targets:
	+    - node-exporter:9100
	```
1. Start the local docker-compose environment 
	```bash
	# starting from the base of deploy-sourcegraph-docker
	cd docker-compose/dev
	./run.sh up -d
	```
1. Open your local grafana at `http://localhost:3370/-/debug/grafana/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Prometheus%22,%7B%22exemplar%22:true%7D%5D`, verify that you're able to see reasonable output for the following queries: 
	- `mount_point_info`
	- `rate(node_disk_reads_completed_total[5m])`
	- `mount_point_info{mount_name="indexDir", instance="zoekt-webserver-0:6070"} * on(device, node_name) group_left rate(node_disk_reads_completed_total[5m])`
1. Grab logs from your local zoekt-webserver instance, you should see output like the following
	```bash
	# starting from the base of deploy-sourcegraph-docker
	cd docker-compose/dev
	> ./run.sh logs zoekt-webserver-0
	+ docker-compose -f docker-compose/docker-compose.yaml -f docker-compose/jaeger/docker-compose.yaml -f docker-compose/dev/docker-compose.yaml logs zoekt-webserver-0
	Attaching to zoekt-webserver-0
	zoekt-webserver-0                | 2022/10/06 23:51:06 INFO: using OpenTelemetry tracer
	zoekt-webserver-0                | 1665100266694356710	DEBUG	zoekt_webserver_mount_info_metrics.deviceNameDiscovery	mountinfo/mountinfo.go:82	github.com/sourcegraph/zoekt/internal/mountinfo.MustRegisterNewMountPointInfoMetric	'stat'-ing mountFilePath	{"Resource": {"service.name": "zoekt-webserver", "service.instance.id": "zoekt-webserver-0:6070"}, "Attributes": {"mountName": "indexDir", "mountFilePath": "/data/index", "operation": "discovering device number"}}
	zoekt-webserver-0                | 1665100266701433710	DEBUG	zoekt_webserver_mount_info_metrics.deviceNameDiscovery	mountinfo/mountinfo.go:103	github.com/sourcegraph/zoekt/internal/mountinfo.MustRegisterNewMountPointInfoMetric	discovered device number	{"Resource": {"service.name": "zoekt-webserver", "service.instance.id": "zoekt-webserver-0:6070"}, "Attributes": {"mountName": "indexDir", "mountFilePath": "/data/index", "operation": "discovering device number", "deviceNumber": "254:1"}}
	zoekt-webserver-0                | 1665100266701895210	DEBUG	zoekt_webserver_mount_info_metrics.deviceNameDiscovery	mountinfo/mountinfo.go:111	github.com/sourcegraph/zoekt/internal/mountinfo.MustRegisterNewMountPointInfoMetric	evaluating sysfs symlink	{"Resource": {"service.name": "zoekt-webserver", "service.instance.id": "zoekt-webserver-0:6070"}, "Attributes": {"mountName": "indexDir", "mountFilePath": "/data/index", "operation": "discovering device path", "symlink": "/sys/dev/block/254:1"}}
	zoekt-webserver-0                | 1665100266702703544	DEBUG	zoekt_webserver_mount_info_metrics.deviceNameDiscovery	mountinfo/mountinfo.go:126	github.com/sourcegraph/zoekt/internal/mountinfo.MustRegisterNewMountPointInfoMetric	discovered device path	{"Resource": {"service.name": "zoekt-webserver", "service.instance.id": "zoekt-webserver-0:6070"}, "Attributes": {"mountName": "indexDir", "mountFilePath": "/data/index", "operation": "discovering device path", "devicePath": "/sys/devices/platform/3f000000.pcie/pci0000:00/0000:00:01.0/virtio1/block/vda/vda1"}}
	zoekt-webserver-0                | 1665100266703207877	DEBUG	zoekt_webserver_mount_info_metrics.deviceNameDiscovery	mountinfo/mountinfo.go:147	github.com/sourcegraph/zoekt/internal/mountinfo.MustRegisterNewMountPointInfoMetric	changing device path since oldDevicePath represents a disk partition	{"Resource": {"service.name": "zoekt-webserver", "service.instance.id": "zoekt-webserver-0:6070"}, "Attributes": {"mountName": "indexDir", "mountFilePath": "/data/index", "operation": "validating device path", "oldDevicePath": "/sys/devices/platform/3f000000.pcie/pci0000:00/0000:00:01.0/virtio1/block/vda/vda1", "newDevicePath": "/sys/devices/platform/3f000000.pcie/pci0000:00/0000:00:01.0/virtio1/block/vda"}}
	zoekt-webserver-0                | 1665100266703671835	DEBUG	zoekt_webserver_mount_info_metrics.deviceNameDiscovery	mountinfo/mountinfo.go:200	github.com/sourcegraph/zoekt/internal/mountinfo.MustRegisterNewMountPointInfoMetric	discovered device name	{"Resource": {"service.name": "zoekt-webserver", "service.instance.id": "zoekt-webserver-0:6070"}, "Attributes": {"mountName": "indexDir", "mountFilePath": "/data/index", "deviceName": "vda"}}
	```

ggilmore committed:
wip: add disk metrics to zoekt

ggilmore committed:
wip

ggilmore committed:
wip

ggilmore committed:
wip

ggilmore committed:
add more logging

ggilmore committed:
wip better comments

ggilmore committed:
comments

ggilmore committed:
remove function

ggilmore committed:
Dockerfile: remove caching

I'll add this back later

ggilmore commented:
Also, should we feature flag this? I'm tempted to say no, since the way that the function is written it'll skip + log directories that are missing / have mismatched permissions.

keegancsmith reviewed (commented):
Very cool stuff! I have some minor stuff inline, but have a suggestion for you since you chatted about testing in our sync.

If you pull out a function were you specify the sysfs path (maybe just as a helper, since as commented it shouldn't be needed by callers) then you can in a test create something that looks like sysfs. The only thing you can't control is what device says for stat on the original path. But in the test you can pick something stat it and then base a fake sysfs based on it.

Additionally I'd add a test on linux that just ensures we find a device for the cwd. IF the test ends up failing in CI then we can remove it. Just hoping this is good enough as an integration test.

ggilmore committed:
remove sysfsMountPoint argument

ggilmore committed:
use only one With statement

ggilmore committed:
extract into sub-function

ggilmore committed:
refactor mount_point_info and proc_metrics to instantiate their own logging scopes

ggilmore committed:
see if CI smoke test works

ggilmore committed:
try verbose

ggilmore committed:
revert test verbose

keegancsmith reviewed (approved):
nice stuff!!

ggilmore committed:
change smoketest to run on all linux targets

ggilmore committed:
change smoketest to run on all linux targets

ggilmore committed:
ttar: add sysfs snapshot folder text fixture

This reuses some tooling that node_exporter uses in their test suite: https://github.com/prometheus/node_exporter/blob/98a40bd712eaeb1fd3ee60ca37e4091c503d8e1c/ttar

This is a real sysfs snapshot taken from the local docker-compose instance that I used for manual testing.

ggilmore committed:
test: abstract out sys mountpoint and deviceNumber fetcher so that we can mock them in tests

ggilmore committed:
test: abstract out sys mountpoint and deviceNumber fetcher so that we can mock them in tests

ggilmore committed:
remove sys from committed fixtures folder

ggilmore committed:
shfmt

ggilmore committed:
ci: add bash and sed to test runner

ggilmore committed:
ci: update ttar script

ggilmore committed:
update ttar

ggilmore committed:
refactor tests to use plain tar

keegancsmith committed:
include lvm.dm-0 snapshot

Updated snapshot script to only capture the directories we are currently
reading. Trying to capture my whole /sys was failing.

ggilmore committed:
tests: convert to table driven tests

ggilmore commented:
@keegancsmith I also sent you this over slack. The snapshot that we took from your machine is incomplete, we also need your `/sys/class` folder.  I removed your test, and I'm gonna merge this as-is. We can follow up with the additional test in a follow up PR.
